File: src/api/contexts.js
import {
  createContext,
  getAliveContext,
  listAliveContexts,
  navigateTo,
  takeScreenshot,
  execAction,
  destroyContext,
} from '../browser/contexts.js';
import { insertLog, getLogsByContext } from '../db/index.js';
import { broadcast } from '../dashboard/sse.js';

export default async function contextRoutes(app) {
  // List all contexts
  app.get('/api/contexts', async () => {
    return listAliveContexts();
  });

  // Create a context
  app.post('/api/contexts', async (req, reply) => {
    const { name, profile, visible } = req.body || {};
    if (!name) return reply.code(400).send({ error: 'name is required' });

    const meta = await createContext({ name, profile, visible });
    insertLog({ contextId: meta.id, level: 'info', message: `Context created: ${name}` });
    broadcast({ event: 'context:created', data: meta });
    reply.code(201).send(meta);
  });

  // Get a single context
  app.get('/api/contexts/:id', async (req, reply) => {
    const entry = getAliveContext(req.params.id);
    if (!entry) return reply.code(404).send({ error: 'Context not found' });
    return entry.meta;
  });

  // Destroy a context
  app.delete('/api/contexts/:id', async (req, reply) => {
    const { id } = req.params;
    try {
      await destroyContext(id);
      insertLog({ contextId: id, level: 'info', message: 'Context destroyed' });
      broadcast({ event: 'context:destroyed', data: { id } });
      return { ok: true };
    } catch (err) {
      return reply.code(404).send({ error: err.message });
    }
  });

  // Navigate to a URL
  app.post('/api/contexts/:id/goto', async (req, reply) => {
    const { url } = req.body || {};
    if (!url) return reply.code(400).send({ error: 'url is required' });

    try {
      const result = await navigateTo(req.params.id, url);
      insertLog({ contextId: req.params.id, level: 'info', message: `Navigated to ${url}` });
      broadcast({ event: 'context:updated', data: { id: req.params.id, url } });
      return result;
    } catch (err) {
      return reply.code(404).send({ error: err.message });
    }
  });

  // Take a screenshot
  app.get('/api/contexts/:id/screenshot', async (req, reply) => {
    try {
      const buffer = await takeScreenshot(req.params.id);
      reply.type('image/png').send(buffer);
    } catch (err) {
      reply.code(404).send({ error: err.message });
    }
  });

  // Execute an action
  app.post('/api/contexts/:id/exec', async (req, reply) => {
    const { action, selector, value, script } = req.body || {};
    if (!action) return reply.code(400).send({ error: 'action is required' });

    try {
      const result = await execAction(req.params.id, { action, selector, value, script });
      insertLog({ contextId: req.params.id, level: 'info', message: `Executed: ${action} ${selector || ''}`.trim() });
      return result;
    } catch (err) {
      return reply.code(400).send({ error: err.message });
    }
  });

  // Get logs for a context
  app.get('/api/contexts/:id/logs', async (req, reply) => {
    try {
      const logs = getLogsByContext(req.params.id);
      return logs;
    } catch (err) {
      return reply.code(404).send({ error: err.message });
    }
  });

  // Append a log entry
  app.post('/api/contexts/:id/logs', async (req, reply) => {
    const { level, message } = req.body || {};
    if (!message) return reply.code(400).send({ error: 'message is required' });

    insertLog({ contextId: req.params.id, level: level || 'info', message });
    broadcast({ event: 'log:new', data: { contextId: req.params.id, level, message } });
    reply.code(201).send({ ok: true });
  });
}

File: src/api/contexts.py
from fastapi import APIRouter, Depends, HTTPException, Request
from fastapi.responses import Response
from pydantic import BaseModel

from .auth import require_token
from ..browser.contexts import (
    create_context,
    get_alive_context,
    list_alive_contexts,
    navigate_to,
    take_screenshot,
    upload_screenshot,
    list_screenshots,
    get_saved_screenshot,
    exec_action,
    get_snapshot,
    destroy_context,
)
from ..db import insert_log, get_logs_by_context
from ..dashboard.sse import broadcast

router = APIRouter(prefix="/api", dependencies=[Depends(require_token)])


class CreateContextBody(BaseModel):
    name: str
    profile: str | None = None
    external: bool = False


class GotoBody(BaseModel):
    url: str
    timeout: int | None = None
    waitUntil: str | None = None


class ExecBody(BaseModel):
    action: str
    selector: str | None = None
    value: str | None = None
    script: str | None = None
    timeout: int | None = None


class LogBody(BaseModel):
    level: str | None = None
    message: str


@router.get("/contexts")
async def list_contexts_route():
    return list_alive_contexts()


@router.post("/contexts", status_code=201)
async def create_context_route(body: CreateContextBody):
    meta = await create_context(name=body.name, profile=body.profile, external=body.external)
    insert_log(context_id=meta["id"], level="info", message=f"Context created: {body.name}")
    broadcast(event="context:created", data=meta)
    return meta


@router.get("/contexts/{ctx_id}")
async def get_context_route(ctx_id: str):
    entry = get_alive_context(ctx_id)
    if not entry:
        raise HTTPException(status_code=404, detail="Context not found")
    return entry["meta"]


@router.delete("/contexts/{ctx_id}")
async def destroy_context_route(ctx_id: str):
    try:
        await destroy_context(ctx_id)
        insert_log(context_id=ctx_id, level="info", message="Context destroyed")
        broadcast(event="context:destroyed", data={"id": ctx_id})
        return {"ok": True}
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/contexts/{ctx_id}/goto")
async def goto_route(ctx_id: str, body: GotoBody):
    try:
        kwargs = {}
        if body.timeout is not None:
            kwargs["timeout"] = body.timeout
        if body.waitUntil is not None:
            kwargs["wait_until"] = body.waitUntil
        result = await navigate_to(ctx_id, body.url, **kwargs)
        insert_log(context_id=ctx_id, level="info", message=f"Navigated to {body.url}")
        broadcast(event="context:updated", data={"id": ctx_id, "url": body.url})
        return result
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/contexts/{ctx_id}/snapshot")
async def snapshot_route(ctx_id: str):
    try:
        return await get_snapshot(ctx_id)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/contexts/{ctx_id}/screenshot")
async def screenshot_route(ctx_id: str):
    try:
        buffer = await take_screenshot(ctx_id)
        return Response(content=buffer, media_type="image/png")
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/contexts/{ctx_id}/screenshot", status_code=204)
async def upload_screenshot_route(ctx_id: str, request: Request):
    body = await request.body()
    if not body:
        raise HTTPException(status_code=400, detail="PNG body required")
    try:
        upload_screenshot(ctx_id, body)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))
    return Response(status_code=204)


@router.get("/contexts/{ctx_id}/screenshots")
async def list_screenshots_route(ctx_id: str):
    return list_screenshots(ctx_id)


@router.get("/contexts/{ctx_id}/screenshots/{filename}")
async def get_screenshot_file_route(ctx_id: str, filename: str):
    try:
        png = get_saved_screenshot(ctx_id, filename)
        return Response(content=png, media_type="image/png")
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.post("/contexts/{ctx_id}/exec")
async def exec_route(ctx_id: str, body: ExecBody):
    try:
        result = await exec_action(
            ctx_id,
            action=body.action,
            selector=body.selector,
            value=body.value,
            script=body.script,
            timeout=body.timeout,
        )
        insert_log(
            context_id=ctx_id,
            level="info",
            message=f"Executed: {body.action} {body.selector or ''}".strip(),
        )
        return result
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/contexts/{ctx_id}/logs")
async def get_logs_route(ctx_id: str):
    return get_logs_by_context(ctx_id)


@router.post("/contexts/{ctx_id}/logs", status_code=201)
async def append_log_route(ctx_id: str, body: LogBody):
    insert_log(context_id=ctx_id, level=body.level or "info", message=body.message)
    broadcast(event="log:new", data={"contextId": ctx_id, "level": body.level, "message": body.message})
    return {"ok": True}
